# Exploring EBay Car Sales Data

This is a guided project done as a part of learning on [Dataquest.io](https://www.dataquest.io/). In this project we would be applying a variety of pandas methods to clean, explore and understand a data set on car listings.

**Following actions/analysis were done on the data as a part of guided project:**

1. Cleaning Column Names
2. Initial exploration and cleaning of data
3. Dealing with Null Data
4. Dealing with incorrect data
5. Finding Outliers in Data
6. Exploring prices by Brand
7. Storing aggregate data in a dataframe

**Apart from what was covered in the guided project, following cleaning/analysis has also been done:**

1. Identifying the categorical columns in the data that uses german words, translate them and map their values to their English counterparts
2. Finding the most common brand-model combinations from the data
3. Th price difference between the cars with some damage and their non-damaged counterparts



## Introduction
In this project, we will be working on a public dataset of used cars from *eBay Kleinanzeigen*, a classifieds section of German ebay website.  
The dataset is available on [Kaggle](https://www.kaggle.com/orgesleka/used-cars-database/data). There are a few modifications done to this dataset, which are as follows:
1. 50,000 data points are sampled from the full dataset
2. The dataset a bit to more closely resemble the real world scraped dataset (the version uploaded to Kaggle was cleaned to be easier to work with)

### Objective
The **aim** of this project is to clean the data and analyze the car listings.

In [331]:
import pandas as pd
import numpy as np
autos = pd.read_csv("autos.csv", encoding='Latin-1') 

In [332]:
autos.info()
autos.head(5)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

**Some Observations:**
1. Dataframe.info()- Out of 20  columns, following 5 have null values:
       1. vehicleType
       2. gearbox
       3. model
       4. fuelType
       5. fuelType
       6. notRepairedDamage
       
None of the above columns have more than **20%** NULL values
The column names use camelcase instead of Python's preferred snakecase,i.e, we can't just replace spaces with underscores

2. Dataframe.head()-
       1. price column is object type but will make more sense being float and can be renamed to priceDollars for more clarity
       2. dateCrawled is a timestamp value
       3. odometer can be renamed to km and its type can be changed to float

In [333]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

** As per the above observations, folllowing name changes will make data in columns more readable: **
1. *yearOfRegistration* to *registration_year* 
2. *monthOfRegistration* to *registration_month*
3. *notRepairedDamage* to *unrepaired_damage*
4. *dateCreated* to *ad_created*
5. camelcase to snakecase

In [334]:
#changing column names

new_columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_PS', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen']
autos.columns = new_columns
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_PS   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In [335]:
autos.describe(include='all')
#autos["price"].head()

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-09 11:54:38  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_PS  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

Following inferences can be taken from the above primary analysis of the dataframe:
1. Columns *seller* and *offer_type* have mostly one value and needs some further analysis
2. Columns *price* and *odometer* should be numeric and data needs to be cleaned weeding out the non-numeric characters


In [336]:
print(autos['seller'].value_counts())
print(autos['offer_type'].value_counts())

privat        49999
gewerblich        1
Name: seller, dtype: int64
Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64


#### Performing cleaning on price and odometer columns
The following operations are performed:
1. Removal of any non-numeric characters
2. Conversion of data type of columns to numeric
3. Renaming of odomter to odometer_km


In [337]:
autos["price"]=autos["price"].str.replace("$","").str.replace(",","").astype(int)
autos["odometer"] = autos["odometer"].str.replace("km","").str.replace(",","").astype(int)
autos.rename(columns = {'odometer':'odometer_km'}, inplace = True)


In [338]:
print(autos["price"].head())
print(autos['odometer_km'].head())

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int64
0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer_km, dtype: int64


**Lets further analyse price and odometer_km columns for any inconsistent/outlier data using the following approach:**
1. Analyse the columns using minimum and maximum values
2. Look for values that seems unrealistically low or high
3. Remove these outliers from our data

The following functions will be used:
1. ***series.unique().shape*** : To see the number of unique values
2. ***series.describe()*** : To view min/max/median etc.
3. ***series.value_counts()*** : To show the different values with their counts
4. ***df[df["col"].between(x,y)]*** : To remove outliers in a column


In [339]:
print("Exploratory Data Analysis for Odometer_Km column:\n") 
eda_odometer_km = autos["odometer_km"].describe() 
print(eda_odometer_km)
print("\nCount of unique values in Odometer_km: " + str(autos["odometer_km"].unique().shape[0]))
print("\nExploratory Data Analysis for Price column:\n") 
eda_price = autos["price"].describe().apply(lambda x: format(x,'f'))
print(eda_price)
print("\nCount of unique values in Price: " + str(autos["price"].unique().shape[0]))

Exploratory Data Analysis for Odometer_Km column:

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

Count of unique values in Odometer_km: 13

Exploratory Data Analysis for Price column:

count       50000.000000
mean         9840.043760
std        481104.380500
min             0.000000
25%          1100.000000
50%          2950.000000
75%          7200.000000
max      99999999.000000
Name: price, dtype: object

Count of unique values in Price: 2357


**As is evident from the Exploratory Data Analysis (EDA) of the data, there might be some outliers present in our data. Let's find and filter those outliers using Inter-Quartile Range** ([IQR](https://en.wikipedia.org/wiki/Interquartile_range#:~:text=In%20descriptive%20statistics%2C%20the%20interquartile,%3D%20Q3%20%E2%88%92%20Q1))

IQR Calculation for *Price* column:


In [340]:
Q1 = float(eda_price["25%"])
Q3 = float(eda_price["75%"])
IQR = Q3 - Q1
autos[autos["price"].between(Q1-1.5*IQR,Q3+1.5*IQR)]["price"].describe()
#print(IQR)
#autos["price"].quantile([0.25,0.5,0.75]) #-- To find quartiles


count    46216.000000
mean      3963.696101
std       3847.238683
min          0.000000
25%       1000.000000
50%       2500.000000
75%       5900.000000
max      16350.000000
Name: price, dtype: float64

IQR Calculation for *Odometer_km* column:

In [341]:
Q1 = float(eda_odometer_km["25%"])
Q3 = float(eda_odometer_km["75%"])
IQR = Q3 - Q1
autos["odometer_km"][autos["odometer_km"].between(Q1-1.5*IQR,Q3+1.5*IQR)].describe()

count     41520.000000
mean     141736.030829
std       17102.004255
min       90000.000000
25%      150000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

As per above observation, if we consider the data points calculated as ouliers using **IQR method**, there is a significant drop in values of mean and standard deviation for price and odometer_km. Let us analyze these data points further. 

In [342]:
print(autos["price"].value_counts().sort_index())
print(autos["odometer_km"].value_counts().sort_index())

0           1421
1            156
2              3
3              1
5              2
8              1
9              1
10             7
11             2
12             3
13             2
14             1
15             2
17             3
18             1
20             4
25             5
29             1
30             7
35             1
40             6
45             4
47             1
49             4
50            49
55             2
59             1
60             9
65             5
66             1
            ... 
151990         1
155000         1
163500         1
163991         1
169000         1
169999         1
175000         1
180000         1
190000         1
194000         1
197000         1
198000         1
220000         1
250000         1
259000         1
265000         1
295000         1
299000         1
345000         1
350000         1
999990         1
999999         2
1234566        1
1300000        1
3890000        1
10000000       1
11111111       2
12345678      

**Observations For Price Column:**

1. Around 30% data points have price 0 dollars, these seems more likely to be outliers.
2. A number of prices is below 30 dollars which might be possible due to the fact that ebay is an action website.
3. There is a significant jump in price after 350000 (around 2X) which seems to be highly improbable. Around 15 prices are more than a million which seems to much.

Hence our outliers would be the data points ie.e prices below or equal to 0 and more than 350000.

**Observations for Odometer_km Column:**

1. More than half of vehicles have 150000 km on their odometer (65%)
2. For this column, the data looks quite reliable i.e, it does not seem to contain any outliers. There is not any sudden increase/decrease found in the data.

Let's remove outliers from our data correspponding to price column

In [343]:
autos = autos[autos["price"].between(1,350000)]
autos["price"].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

In [344]:
autos[["date_crawled","last_seen","ad_created"]].head()

date_crawled            last_seen           ad_created
0  2016-03-26 17:47:46  2016-04-06 06:45:54  2016-03-26 00:00:00
1  2016-04-04 13:38:56  2016-04-06 14:45:08  2016-04-04 00:00:00
2  2016-03-26 18:57:24  2016-04-06 20:15:37  2016-03-26 00:00:00
3  2016-03-12 16:58:10  2016-03-15 03:16:28  2016-03-12 00:00:00
4  2016-04-01 14:38:50  2016-04-01 14:38:50  2016-04-01 00:00:00

**Observation**:

By seeing the values in **date_crawled, last_seen** and **ad_created**, we can infer that the first 10 characters of each record contains the date. Hence, to understand the date range, we can extract just the date values, use *Series.value_counts()* to generate a distribution, and then sort by the index.

In [345]:
date_crawled_distribution = (autos["date_crawled"]
                             .str[:10].value_counts(normalize = True, dropna = False)
                             .sort_index())
date_crawled_distribution

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

**Findings**

1. The data was regularly crawled for about a month from 5th Mar to 7th Apr, 2016.
2. We can see that there is a uniformity in the distribution of the ads crawled daily.

In [346]:
ad_created_distribution = (autos["ad_created"].str[:10]
                           .value_counts(normalize = True, dropna = False)
                           .sort_index())
print(ad_created_distribution.count())
ad_created_distribution

76


2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000041
2016-01-13    0.000021
2016-01-14    0.000021
2016-01-16    0.000021
2016-01-22    0.000021
2016-01-27    0.000062
2016-01-29    0.000021
2016-02-01    0.000021
2016-02-02    0.000041
2016-02-05    0.000041
2016-02-07    0.000021
2016-02-08    0.000021
2016-02-09    0.000021
2016-02-11    0.000021
2016-02-12    0.000041
2016-02-14    0.000041
2016-02-16    0.000021
2016-02-17    0.000021
2016-02-18    0.000041
2016-02-19    0.000062
2016-02-20    0.000041
2016-02-21    0.000062
                ...   
2016-03-09    0.033151
2016-03-10    0.031895
2016-03-11    0.032904
2016-03-12    0.036755
2016-03-13    0.017008
2016-03-14    0.035190
2016-03-15    0.034016
2016-03-16    0.030125
2016-03-17    0.031278
2016-03-18    0.013590
2016-03-19    0.033687
2016-03-20    0.037949
2016-03-21 

In [347]:
(ad_created_distribution[ad_created_distribution.keys()
                         > '2016-03-04']).sort_index()

2016-03-05    0.022897
2016-03-06    0.015320
2016-03-07    0.034737
2016-03-08    0.033316
2016-03-09    0.033151
2016-03-10    0.031895
2016-03-11    0.032904
2016-03-12    0.036755
2016-03-13    0.017008
2016-03-14    0.035190
2016-03-15    0.034016
2016-03-16    0.030125
2016-03-17    0.031278
2016-03-18    0.013590
2016-03-19    0.033687
2016-03-20    0.037949
2016-03-21    0.037579
2016-03-22    0.032801
2016-03-23    0.032060
2016-03-24    0.029280
2016-03-25    0.031751
2016-03-26    0.032266
2016-03-27    0.030989
2016-03-28    0.034984
2016-03-29    0.034037
2016-03-30    0.033501
2016-03-31    0.031875
2016-04-01    0.033687
2016-04-02    0.035149
2016-04-03    0.038855
2016-04-04    0.036858
2016-04-05    0.011819
2016-04-06    0.003253
2016-04-07    0.001256
Name: ad_created, dtype: float64

In [348]:
(ad_created_distribution[ad_created_distribution.keys()
                         <= '2016-03-04']).sort_index()

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000041
2016-01-13    0.000021
2016-01-14    0.000021
2016-01-16    0.000021
2016-01-22    0.000021
2016-01-27    0.000062
2016-01-29    0.000021
2016-02-01    0.000021
2016-02-02    0.000041
2016-02-05    0.000041
2016-02-07    0.000021
2016-02-08    0.000021
2016-02-09    0.000021
2016-02-11    0.000021
2016-02-12    0.000041
2016-02-14    0.000041
2016-02-16    0.000021
2016-02-17    0.000021
2016-02-18    0.000041
2016-02-19    0.000062
2016-02-20    0.000041
2016-02-21    0.000062
2016-02-22    0.000021
2016-02-23    0.000082
2016-02-24    0.000041
2016-02-25    0.000062
2016-02-26    0.000041
2016-02-27    0.000124
2016-02-28    0.000206
2016-02-29    0.000165
2016-03-01    0.000103
2016-03-02    0.000103
2016-03-03    0.000865
2016-03-04    0.001483
Name: ad_created, dtype: float64

We can see that around **45% **ads were created within a similar time frame as the crawling dates, with ads to date distribution ranging between **0.12% to 3.88% **. But, some of the ads were created much older, with ads to date distribution ranging between **0.002% to 1.4%**.
It can be inferred from the above information that very few listings for ads were created between 11th June 2015 - 4th March 2016

In [349]:
(autos["last_seen"].str[:10]
 .value_counts(normalize = True, dropna = False)
 .sort_index())

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64

Observations:

This last_seen column tells us of when an ad has been delisted from the site, most likely due to a sale.
There is a spike of distribution for the last three days which seems unlikely as the days prior to this have roughly similar distribution.

Now let's analyze the registration year column.

In [350]:
autos["registration_year"].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

As is evident from the above statististics, the minimum and maximum registration years of 1000 and 9999 respectively do not make sense.

Let's find out the number of registrations in each year.

In [351]:
autos["registration_year"].value_counts().sort_index()

1000       1
1001       1
1111       1
1800       2
1910       5
1927       1
1929       1
1931       1
1934       2
1937       4
1938       1
1939       1
1941       2
1943       1
1948       1
1950       3
1951       2
1952       1
1953       1
1954       2
1955       2
1956       4
1957       2
1958       4
1959       6
1960      23
1961       6
1962       4
1963       8
1964      12
        ... 
2000    3156
2001    2636
2002    2486
2003    2699
2004    2703
2005    2936
2006    2670
2007    2277
2008    2215
2009    2085
2010    1589
2011    1623
2012    1310
2013     803
2014     663
2015     392
2016    1220
2017    1392
2018     470
2019       2
2800       1
4100       1
4500       1
4800       1
5000       4
5911       1
6200       1
8888       1
9000       1
9999       3
Name: registration_year, Length: 95, dtype: int64

Observations:
    
A car that was listed in 2016 cannot have a registration year greater than 2016. While analyzing the Last seen column, we saw that the maximum value of year is 2016. Hence, any entry of registration year after 2016 is definitely inaccurate.

Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s.

Let's count the number of listings with cars that fall outside the 1900 - 2016 interval and see if it's safe to remove those rows entirely, or if we need more custom logic.

In [352]:
(autos["registration_year"].count()-
 autos[autos["registration_year"].between(1900,2016)]["registration_year"].count())


1884

Observation:

About **4%** of the total registrations were done outside the range of 1900-2016. So  we can safely remove these rows entirely (since they amount to very less percentage of data).

Let's remove the values outside those upper and lower bounds and calculate the distribution of the remaining values using *Series.value_counts(normalize=True)*

In [353]:
autos = autos[autos["registration_year"].between(1900,2016)]

#Finding the normalized values for cleaned data
norm_cnts_reg_yr = autos["registration_year"].value_counts(normalize = True).sort_index()
norm_cnts_reg_yr

1910    0.000107
1927    0.000021
1929    0.000021
1931    0.000021
1934    0.000043
1937    0.000086
1938    0.000021
1939    0.000021
1941    0.000043
1943    0.000021
1948    0.000021
1950    0.000064
1951    0.000043
1952    0.000021
1953    0.000021
1954    0.000043
1955    0.000043
1956    0.000086
1957    0.000043
1958    0.000086
1959    0.000129
1960    0.000493
1961    0.000129
1962    0.000086
1963    0.000171
1964    0.000257
1965    0.000364
1966    0.000471
1967    0.000557
1968    0.000557
          ...   
1987    0.001542
1988    0.002892
1989    0.003727
1990    0.007433
1991    0.007262
1992    0.007926
1993    0.009104
1994    0.013474
1995    0.026285
1996    0.029412
1997    0.041794
1998    0.050620
1999    0.062060
2000    0.067608
2001    0.056468
2002    0.053255
2003    0.057818
2004    0.057904
2005    0.062895
2006    0.057197
2007    0.048778
2008    0.047450
2009    0.044665
2010    0.034040
2011    0.034768
2012    0.028063
2013    0.017202
2014    0.0142

It is clear from the above data that most of the cars were registered between 1994-2016 during which the registration spiked from 1997-2005. After 2005 though, there has been a steady decline. 2015 stands out especially because it has the least number of car registrations in the first decade of 21st century.

In [354]:
print(norm_cnts_reg_yr.count())

#Finding normalized distribution of registration year where less than 1% cars were registered
less_than_1 = norm_cnts_reg_yr[norm_cnts_reg_yr < 0.01]
less_than_1.sort_index()

78


1910    0.000107
1927    0.000021
1929    0.000021
1931    0.000021
1934    0.000043
1937    0.000086
1938    0.000021
1939    0.000021
1941    0.000043
1943    0.000021
1948    0.000021
1950    0.000064
1951    0.000043
1952    0.000021
1953    0.000021
1954    0.000043
1955    0.000043
1956    0.000086
1957    0.000043
1958    0.000086
1959    0.000129
1960    0.000493
1961    0.000129
1962    0.000086
1963    0.000171
1964    0.000257
1965    0.000364
1966    0.000471
1967    0.000557
1968    0.000557
1969    0.000407
1970    0.000814
1971    0.000557
1972    0.000707
1973    0.000493
1974    0.000514
1975    0.000386
1976    0.000450
1977    0.000471
1978    0.000943
1979    0.000728
1980    0.001821
1981    0.000600
1982    0.000878
1983    0.001093
1984    0.001093
1985    0.002035
1986    0.001542
1987    0.001542
1988    0.002892
1989    0.003727
1990    0.007433
1991    0.007262
1992    0.007926
1993    0.009104
2015    0.008397
Name: registration_year, dtype: float64

The analysis above shows the years for which the car registrations were less than 1%. Majority of the years belong to 1900s which is understadable as during those times, the market for cars was very small.

In [355]:
unique_brands = autos["brand"].value_counts()
unique_brands.count()

40

There are **40** unique brands of cars which are listed on the Ebay website. Let's analyze the share of the listings for various car brands on the website.

In [356]:
#finding share distribution in the listings for various car brands 
normalized_brand_data = (autos["brand"]
                         .value_counts(normalize = True).sort_values())
normalized_brand_data

lada              0.000578
lancia            0.001071
rover             0.001328
trabant           0.001392
daewoo            0.001500
jaguar            0.001564
saab              0.001649
land_rover        0.002099
subaru            0.002142
jeep              0.002271
daihatsu          0.002506
dacia             0.002635
chrysler          0.003513
chevrolet         0.005698
suzuki            0.005934
porsche           0.006127
alfa_romeo        0.006641
kia               0.007069
honda             0.007840
mitsubishi        0.008226
mini              0.008762
volvo             0.009147
sonstige_autos    0.009811
hyundai           0.010025
toyota            0.012703
citroen           0.014010
smart             0.014160
mazda             0.015188
nissan            0.015274
skoda             0.016409
seat              0.018273
fiat              0.025642
peugeot           0.029841
renault           0.047150
ford              0.069900
audi              0.086566
mercedes_benz     0.096463
o

Now, let us find out the car brands which have more than **5%** share in the total listings on the website.

In [357]:
#finding brands which have more than 5% share of the total car brands
chosen_brands = normalized_brand_data[normalized_brand_data*100 >5]*100
chosen_brands

ford              6.989996
audi              8.656627
mercedes_benz     9.646323
opel             10.758124
bmw              11.004477
volkswagen       21.126368
Name: brand, dtype: float64

It is lucid from the above analysis that following 6 major brands control more than 5% of the listings on german auction website:

1. *Ford* with *6.9%* listings
2. *Audi* with *8.6%* listings
3. *Mercedes Benz* with *9.6%* listings
4. *Opel* with *10.7%* listings
5. *BMW* with *11.0%* listings
6. *Volkswagen* with around *21.0%* listings

Clearly, **Volkwagen** has the highest percentage of listings (about 10% more than the first runner up) among the top brands.
Let us analyse the mean price of car listed for each brand.

In [358]:
#import operator
#autos.head()
sales_agg = {}
for brand in chosen_brands.index:
    brand_price_data = autos[autos["brand"]==brand]["price"]
    mean_price = brand_price_data.mean()
    sales_agg[brand] = int(mean_price)
    
#print(sorted(sales_agg.items(), key = operator.itemgetter(1)))
sales_agg

{'audi': 9336,
 'bmw': 8332,
 'ford': 3749,
 'mercedes_benz': 8628,
 'opel': 2975,
 'volkswagen': 5402}

We can see that there is a considerable price difference between the competitors. Audi, BMW and Mercedes, all have their mean price in tha range of 8000\$- 9500\$ while the leading brand, Volkswagen, has its mean price in the midrange of 5402\$. Opel has the cheapest cars. The midrange price of Volkswagen is proably the reason why its cars have more listings on Ebay.

Let us find out the average mileage for these brands.

In [359]:
mileage = {}

for brand in chosen_brands.index:
    distance = autos[autos["brand"] == brand]["odometer_km"]
    avg_distance = distance.mean()
    mileage[brand] = int(avg_distance)
    
mileage

{'audi': 129157,
 'bmw': 132572,
 'ford': 124266,
 'mercedes_benz': 130788,
 'opel': 129310,
 'volkswagen': 128707}

In [360]:
#Converting mileage and sales_agg into series objects

mileage_series = pd.Series(mileage).sort_values(ascending = False)
print(mileage_series)
print("\n")
mean_price_series = pd.Series(sales_agg).sort_values(ascending = False)
print(mean_price_series)
market_share = pd.Series(chosen_brands).astype(int)
print("\n")
print(market_share)

bmw              132572
mercedes_benz    130788
opel             129310
audi             129157
volkswagen       128707
ford             124266
dtype: int64


audi             9336
mercedes_benz    8628
bmw              8332
volkswagen       5402
ford             3749
opel             2975
dtype: int64


ford              6
audi              8
mercedes_benz     9
opel             10
bmw              11
volkswagen       21
Name: brand, dtype: int64


In [361]:
#Creating a dataframe containg mileage and mean_price of top brands for comparison

mileage_mean_price_df = pd.DataFrame(mean_price_series,
                                    columns = ["mean_price"])
mileage_mean_price_df["mileage"] = mileage_series
mileage_mean_price_df["market_share"] = market_share
mileage_mean_price_df.sort_values(by=["market_share"], ascending = False)

mean_price  mileage  market_share
volkswagen           5402   128707            21
bmw                  8332   132572            11
opel                 2975   129310            10
mercedes_benz        8628   130788             9
audi                 9336   129157             8
ford                 3749   124266             6

# Conclusion:

The above dataframe shows the analysis of mean price, mileage and market share.There is not much difference in the mileage of the cars.

We can see that even though Ford cars have the lowest mean price, their mileage is also the lowest, which is probably the cause behind its lowest market share.

The situation becomes interesting when we see the case of top three brands. 
Although Opel cars have lower mean price and better mileage than Volkwagen cars, they are still behind Volkwagen by 11%.

# Extra Fun Stuff!!

## Let us clean some more data!

Firstly, there are many categorical columns whose values contains data in german. For more uniformness, let's translate that data into English.


In [362]:
autos._get_numeric_data().columns # to find numeric columns

autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type  price   abtest vehicle_type  registration_year  \
0  privat    Angebot   5000  control          bus               2004   
1  privat    Angebot   8500  control    limousine               1997   
2  privat    Angebot   8990     test    limousine               2009   
3  privat    Angebot   4350  control   kleinwagen               2007   
4  privat    Angebot   1350     test        kombi               2003   

     gearbox  power_PS   model  odometer_km  registration_month fuel_type  \
0    manuell       158  andere       150000                   3       lpg   
1  automatik       286     7er       150000                   6    benzin   
2    manuell       102    golf        70000                   7    benzin   
3  automatik        71  fortwo        70000                   6    benzin   
4    manuell         0   focus       150000                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In [363]:
autos["seller"].value_counts()
seller_mapping = {'privat':'private','gewerblich':'commercial'}

In [364]:
autos["seller"] = autos["seller"].map(seller_mapping)
autos["seller"].value_counts()

private       46680
commercial        1
Name: seller, dtype: int64

In [365]:
autos["offer_type"].value_counts()
offer_type_mapping = {'Angebot':'Offer'}
autos["offer_type"].value_counts()

Angebot    46681
Name: offer_type, dtype: int64

In [366]:
autos["offer_type"] = autos["offer_type"].map(offer_type_mapping)
autos["offer_type"].value_counts()

Offer    46681
Name: offer_type, dtype: int64

In [367]:
autos["vehicle_type"].value_counts()
'''vehice_type_mapping = {'kleinwagen':'small car',
                       'kombi':'station wagon',
                       'cabrio':'convertible',
                       'andere':'other'}'''
#mapping does not work by replacing only the mapped variables,
#it will keep only mapped variables when assigning mapping to resp. column
autos["vehicle_type"].value_counts()

limousine     12598
kleinwagen    10585
kombi          8930
bus            4031
cabrio         3016
coupe          2462
suv            1965
andere          390
Name: vehicle_type, dtype: int64

In [368]:
autos["vehicle_type"] = (autos["vehicle_type"]
                         .replace(['kleinwagen','kombi','cabrio','andere'],
                            ['small car','station wagon','convertible','other']
                                 ))
autos["vehicle_type"].value_counts()

limousine        12598
small car        10585
station wagon     8930
bus               4031
convertible       3016
coupe             2462
suv               1965
other              390
Name: vehicle_type, dtype: int64

In [369]:
autos["gearbox"].value_counts()
gearbox_mapping = {'manuell':'manual','automatik':'automatic'}
autos["gearbox"].value_counts()

manuell      34715
automatik     9856
Name: gearbox, dtype: int64

In [370]:
autos["gearbox"] = autos["gearbox"].map(gearbox_mapping)
autos["gearbox"].value_counts()

manual       34715
automatic     9856
Name: gearbox, dtype: int64

In [371]:
autos["model"].value_counts()
autos["model"] = autos["model"].replace(['andere'],['other'])
autos["model"].value_counts()

golf                  3707
other                 3373
3er                   2615
polo                  1609
corsa                 1592
passat                1349
astra                 1348
a4                    1231
c_klasse              1136
5er                   1132
e_klasse               958
a3                     825
a6                     797
focus                  762
fiesta                 722
transporter            674
twingo                 615
2_reihe                600
fortwo                 550
vectra                 544
a_klasse               539
1er                    521
3_reihe                486
mondeo                 479
clio                   473
touran                 433
punto                  415
zafira                 394
ka                     349
megane                 335
                      ... 
sirion                  11
terios                  10
300c                    10
gl                      10
mii                     10
lybra                   10
l

In [372]:
autos["fuel_type"].value_counts()

benzin     28540
diesel     14032
lpg          649
cng           71
hybrid        37
elektro       19
andere        15
Name: fuel_type, dtype: int64

In [373]:
autos["fuel_type"] = (autos["fuel_type"]
                      .replace(['benzin','elektro','andere'],
                               ['gasoline','electrical','other']))
autos["fuel_type"].value_counts()

gasoline      28540
diesel        14032
lpg             649
cng              71
hybrid           37
electrical       19
other            15
Name: fuel_type, dtype: int64

In [376]:
autos["unrepaired_damage"].value_counts()

nein    33834
ja       4540
Name: unrepaired_damage, dtype: int64

In [379]:
unrepaired_damage = {
    'nein': 'no',
    'ja': 'yes'
}
autos['unrepaired_damage'] = autos['unrepaired_damage'].map(unrepaired_damage)
autos["unrepaired_damage"].value_counts()

no     33834
yes     4540
Name: unrepaired_damage, dtype: int64

## Let's do some more analysis !

We are interested to find out the **Blockbuster Models** (model having maximum listings) for brands having listings more than 5%.

In [390]:
brand_model = {}

for brand in chosen_brands.index:
    blockbuster_model = autos[autos["brand"] == brand]["model"].value_counts(ascending = False).index[0]
    brand_model[brand] = blockbuster_model
    
brand_model

{'audi': 'a4',
 'bmw': '3er',
 'ford': 'focus',
 'mercedes_benz': 'c_klasse',
 'opel': 'corsa',
 'volkswagen': 'golf'}

Now, we are interested to find out how much price difference is there between the cars with some damage and their non-damaged counterparts.

In [392]:
damaged_yes_no = ['no','yes']
damaged_price = {}
for value in damaged_yes_no:
    price = autos[autos["unrepaired_damage"] == value]["price"] #returns a series
    mean_price = price.mean()
    damaged_price[value] = int(mean_price)
    
damaged_price

{'no': 7164, 'yes': 2241}

From the above analysis, it can be concluded that cars with damage were listed for way cheaper price than their non-damaged counterparts. The price difference is about 